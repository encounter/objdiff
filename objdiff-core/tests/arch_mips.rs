use objdiff_core::{diff, obj};

mod common;

#[test]
#[cfg(feature = "mips")]
fn read_mips() {
    let diff_config = diff::DiffObjConfig { mips_register_prefix: true, ..Default::default() };
    let obj =
        obj::read::parse(include_object!("data/mips/main.c.o"), &diff_config, obj::DiffSide::Base)
            .unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "ControlEntry").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "mips")]
fn cross_endian_diff() {
    let diff_config = diff::DiffObjConfig::default();
    let obj_be =
        obj::read::parse(include_object!("data/mips/code_be.o"), &diff_config, obj::DiffSide::Base)
            .unwrap();
    assert_eq!(obj_be.endianness, object::Endianness::Big);
    let obj_le =
        obj::read::parse(include_object!("data/mips/code_le.o"), &diff_config, obj::DiffSide::Base)
            .unwrap();
    assert_eq!(obj_le.endianness, object::Endianness::Little);
    let left_symbol_idx = obj_be.symbols.iter().position(|s| s.name == "func_00000000").unwrap();
    let right_symbol_idx =
        obj_le.symbols.iter().position(|s| s.name == "func_00000000__FPcPc").unwrap();
    let (left_diff, right_diff) =
        diff::code::diff_code(&obj_be, &obj_le, left_symbol_idx, right_symbol_idx, &diff_config)
            .unwrap();
    // Although the objects differ in endianness, the instructions should match.
    assert_eq!(left_diff.instruction_rows[0].kind, diff::InstructionDiffKind::None);
    assert_eq!(right_diff.instruction_rows[0].kind, diff::InstructionDiffKind::None);
    assert_eq!(left_diff.instruction_rows[1].kind, diff::InstructionDiffKind::None);
    assert_eq!(right_diff.instruction_rows[1].kind, diff::InstructionDiffKind::None);
    assert_eq!(left_diff.instruction_rows[2].kind, diff::InstructionDiffKind::None);
    assert_eq!(right_diff.instruction_rows[2].kind, diff::InstructionDiffKind::None);
}

#[test]
#[cfg(feature = "mips")]
fn filter_non_matching() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(
        include_object!("data/mips/vw_main.c.o"),
        &diff_config,
        obj::DiffSide::Base,
    )
    .unwrap();
    insta::assert_debug_snapshot!(obj.symbols);
}
