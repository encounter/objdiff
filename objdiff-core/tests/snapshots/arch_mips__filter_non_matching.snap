---
source: objdiff-core/tests/arch_mips.rs
expression: obj.symbols
---
[
    Symbol {
        name: "build/src/bodyprog/view/vw_main.i",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.text]",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Section,
        section: Some(
            0,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.data]",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Section,
        section: Some(
            2,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.bss]",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Section,
        section: Some(
            3,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "gcc2_compiled.",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: Some(
            0,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "__gnu_compiled_c",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: Some(
            0,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: ".endfunc_80048AF4",
        demangled_name: None,
        address: 424,
        size: 0,
        kind: Unknown,
        section: Some(
            0,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: ".endfunc_80048DA8",
        demangled_name: None,
        address: 1028,
        size: 0,
        kind: Unknown,
        section: Some(
            0,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: ".endfunc_80048E3C",
        demangled_name: None,
        address: 1264,
        size: 0,
        kind: Unknown,
        section: Some(
            0,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.reginfo]",
        demangled_name: None,
        address: 0,
        size: 24,
        kind: Section,
        section: Some(
            4,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.MIPS.abiflags]",
        demangled_name: None,
        address: 0,
        size: 24,
        kind: Section,
        section: Some(
            5,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.pdr]",
        demangled_name: None,
        address: 0,
        size: 320,
        kind: Section,
        section: Some(
            6,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.gnu.attributes]",
        demangled_name: None,
        address: 0,
        size: 16,
        kind: Section,
        section: Some(
            8,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwInitViewInfo",
        demangled_name: None,
        address: 0,
        size: 88,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwViewPointInfo",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "GsInitCoordinate2",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwSetViewInfo",
        demangled_name: None,
        address: 784,
        size: 96,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwGetViewCoord",
        demangled_name: None,
        address: 88,
        size: 12,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwGetViewPosition",
        demangled_name: None,
        address: 100,
        size: 40,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwGetViewAngle",
        demangled_name: None,
        address: 140,
        size: 48,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80048AF4",
        demangled_name: None,
        address: 188,
        size: 236,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global | Ignored),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "ratan2",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "SquareRoot0",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80096C94",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwSetViewInfoDirectMatrix",
        demangled_name: None,
        address: 696,
        size: 88,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80048AF4.NON_MATCHING",
        demangled_name: None,
        address: 188,
        size: 236,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global | Ignored),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwSetCoordRefAndEntou",
        demangled_name: None,
        address: 424,
        size: 272,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80096E78",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "shRsin",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "shRcos",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vbSetRefView",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "vwMatrixToAngleYXZ",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Unknown,
        section: None,
        flags: FlagSet(Global),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80048DA8",
        demangled_name: None,
        address: 880,
        size: 148,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global | Ignored),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80048DA8.NON_MATCHING",
        demangled_name: None,
        address: 880,
        size: 148,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global | Ignored),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80048E3C",
        demangled_name: None,
        address: 1028,
        size: 236,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global | Ignored),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "func_80048E3C.NON_MATCHING",
        demangled_name: None,
        address: 1028,
        size: 236,
        kind: Function,
        section: Some(
            0,
        ),
        flags: FlagSet(Global | Ignored),
        align: None,
        virtual_address: None,
    },
    Symbol {
        name: "[.data-0]",
        demangled_name: None,
        address: 0,
        size: 0,
        kind: Section,
        section: Some(
            2,
        ),
        flags: FlagSet(Local),
        align: None,
        virtual_address: None,
    },
]
